projects/illm/gan_compression_module.py

```diff
@@ -0,0 +1,309 @@
+        latent_projector: Optional[nn.Module] = None,
+        generator_weight: float = 1.0,
+        gradient_clip_val: Optional[float] = None,
+        gradient_clip_algorithm: str = "norm",
+        noisy_context_steps: int = 0,
+        mc_sampling: bool = True,
+    ):
+        super().__init__(
+            model=model,
+            target_rate_config=target_rate_config,
+            optimizer_config=optimizer_config,
+            distortion_lam=distortion_lam,
+            distortion_loss=distortion_loss,
+            gradient_clip_val=gradient_clip_val,
+            gradient_clip_algorithm=gradient_clip_algorithm,
+            noisy_context_steps=noisy_context_steps,
+        )
+
+        self.discriminator = discriminator
+        self.discriminator_loss = discriminator_loss
+        self.latent_projector = latent_projector
+        self.generator_loss = generator_loss
+        self.generator_weight = generator_weight
+        self.mc_sampling = mc_sampling
+
+    def disable_gradients(self, opt):
+        for group in opt.param_groups:
+            for param in group["params"]:
+                param.requires_grad_(False)
+
+    def enable_gradients(self, opt):
+        for group in opt.param_groups:
+            for param in group["params"]:
+                param.requires_grad_(True)
+
+    def run_discriminator(self, images: Tensor, context: Tensor) -> Tensor:
+        if self.discriminator.is_conditional:
+            return self.discriminator(images, context)
+        else:
+            return self.discriminator(images)
+
+    def forward(self, x):
+        return self.model(x)
+
+    def training_step(self, batch, batch_idx):
+        full_images: Tensor
+        output: HyperpriorOutput
+        opts = self.optimizers()
+        scheds = self.lr_schedulers()
+        if len(opts) == 3:
+            model_opt, disc_opt, quantile_opt = opts
+            model_sched, disc_sched, quantile_sched = scheds
+        else:
+            model_opt, disc_opt = opts
+            model_sched, disc_sched = scheds
+            quantile_opt = None
+
+        # if doing Monte Carlo sampling, split the batch for the discriminator
+        full_images, _ = batch
+        if self.mc_sampling is True:
+            split_ind = full_images.shape[0] // 2
+            model_images = full_images[:split_ind]
+            disc_images = full_images[split_ind : split_ind * 2]
+        else:
+            model_images = full_images
+            disc_images = full_images
+
+        # get the discriminator/generator targets
+        if self.latent_projector is not None:
+            with torch.no_grad():
+                fake_target = self.latent_projector(model_images)
+                real_target = self.latent_projector(disc_images)
+        else:
+            real_target = torch.tensor(1, dtype=torch.long)
+            fake_target = torch.tensor(1, dtype=torch.long)
+
+        ########### GENERATOR STEP ###########
```

Comment:
same as above: could extract a function named - generator step

Reply:
I worked on this a little bit, but opted against it for a couple of reasons:

1. There are a bunch of inter-related variables, so we don't save very much by functionalizing.
2. Philosophically I think the training step should give a very good picture of the global state for calculating the loss, and to give this picture distribution should be avoided.